plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Activator.java

```diff
@@ -47,9 +47,9 @@ private void visit(IJavaElementDelta delta) {
         case IJavaElement.JAVA_PROJECT:
           if ((delta.getFlags() & IJavaElementDelta.F_CLASSPATH_CHANGED) != 0) {
             final IJavaProject javaProject = (IJavaProject) delta.getElement();
-            Job updateContainerStateJob = new WorkspaceJob("Updating Google Cloud libraries") {
+            Job updateContainerStateJob = new UIJob("Updating Google Cloud libraries") {
               @Override
-              public IStatus runInWorkspace(IProgressMonitor monitor) {
+              public IStatus runInUIThread(IProgressMonitor monitor) {
```

Comment:
This is a very surprising fix. How does this job interact with the UI?

Reply:
The gist of this change is to make the code run in the main UI thread, instead of a separate thread that runs concurrently. It doesn't update any UI. (We actually mark the job as a system job (`setSystem(true)`) to make it invisible to users.) I believe usual classpath updates except for this update happen only in the main UI thread (at least for our code), so this eliminates the concurrent classpath updates that cause the problem.

Reply:
We can wait what @briandealwis will say.

Reply:
Is it possible the problem is that those other jobs are running on the UI thread and shouldn't be?

Reply:
As long as there is any concurrent "job" updating our library container without properly synchronization, the problem does not go away. So making a job not run in the UI thread would actually aggravate the problem.

But now you said it, I'm kind of losing faith in whether all of our updates are currently being done in the UI thread except for this job. I initially thought so, but we need to verify this. If, at the current state, there are already other jobs running in a non-UI thread, then this problem is really messy to make it right. The code pieces to update our library container is spread too widely and in complicated or inter-class ways in the code base, and making each of such updates transactionally atomic with proper synchronization may turn into quite a big project.

Reply:
Sorry for this late reply.  Classpath changes can happen on any thread, and an operation on the UI thread can be interrupted should anything trigger an SWT event dispatch.  So I think we're better off *not* doing anything on the UI thread and instead securing the update job with the appropriate scheduling rules.  Which in this case is the workspace's rule-factory's build rule, from what I can see (`getWorkspace().getRuleFactory().buildRule()`).